### ISSUE TITLE
bug: spurious "partially unknown"

### ISSUE THREAD

**Issue by bukzor** [microsoft/pyright#10850]
**Describe the bug**

I don't understand this bug at all, but I can tell you how to reproduce it, and I think it's clearly a bug.

Setup:

```console
$ tail -vn99 demo*
==> demo_typed.py <==
from __future__ import annotations

from typing import TypeVar

J = TypeVar("J", bound="JSON")
JSONObjectOf = dict[str, J]
JSON = str | JSONObjectOf["JSON"]
JSONObject = JSONObjectOf[JSON]

==> demo.py <==
from demo_typed import JSON
from demo_typed import JSONObjectOf


def identity(json: JSONObjectOf[JSON]):
    return json

==> demo0.py <==
import demo_typed


def identity0(json0: demo_typed.JSONObjectOf[demo_typed.JSON]):
    return json0

==> demo1.py <==
from demo_typed import JSON
from demo_typed import JSONObjectOf


def identity1(json1: JSONObjectOf[JSON]):
    return json1

==> demo2.py <==
from demo_typed import JSONObject


def identity2(json2: JSONObject):
    return json2

==> demo3.py <==
import demo_typed


def identity3(json3: demo_typed.JSONObject):
    return json3
```

**expected behavior**:

```
$ pyright demo0.py demo2.py | sed 's#'"$PWD/"'#./#g'
0 errors, 0 warnings, 0 informations
```

**actual behavior**:

```
$ pyright demo2.py demo0.py | sed 's#'"$PWD/"'#./#g'
./demo0.py
  ./demo0.py:4:5 - error: Return type, "dict[str, str | Unknown]", is partially unknown (reportUnknownParameterType)
  ./demo0.py:4:15 - error: Type of parameter "json0" is partially unknown
    Parameter type is "dict[str, str | Unknown]" (reportUnknownParameterType)
  ./demo0.py:4:46 - error: Type of "JSON" is partially unknown
    Type of "JSON" is "type[str] | Unknown" (reportUnknownMemberType)
  ./demo0.py:5:12 - error: Return type, "dict[str, str | Unknown]", is partially unknown (reportUnknownVariableType)
./demo2.py
  ./demo2.py:4:5 - error: Return type, "dict[str, str | Unknown]", is partially unknown (reportUnknownParameterType)
  ./demo2.py:4:15 - error: Type of parameter "json2" is partially unknown
    Parameter type is "dict[str, str | Unknown]" (reportUnknownParameterType)
  ./demo2.py:5:12 - error: Return type, "dict[str, str | Unknown]", is partially unknown (reportUnknownVariableType)
7 errors, 0 warnings, 0 informations
```

NOTE:

* the only difference between the commands is the ordering of files.
* the only difference between files is the details of *how* objects are imported.

**VS Code extension or command-line**

```
$ pyright --version
pyright 1.1.404

$ cat pyproject.toml
[tool.pyright]
reportUnknownParameterType = "error"
```

**Comment by erictraut** [comment]
Hmm, I'm not able to repro the problem as described. I don't see any errors regardless of the file order. Can you think of any other steps to repro?

Do you suspect that this is a regression? Are you able to repro it with older versions?

Does it require a particular pythonVersion?

**Comment by pinterior** [comment]
I could repro the problem as following:

### Files

- typ.py
```py
Nested = list["StrOrNested"]
StrOrNested = str | Nested
Alias = Nested
```
- a.py
```py
# pyright: strict

from typ import Alias

def identity(x: Alias):
    return x

reveal_type(identity)
```
- b.py
```py
# pyright: strict

from typ import Nested

def identity(x: Nested):
    return x

reveal_type(identity)
```

### Run

```
$ pyright a.py
a.py
  a.py:5:5 - error: Return type, "list[str | Unknown]", is partially unknown (reportUnknownParameterType)
  a.py:5:14 - error: Type of parameter "x" is partially unknown
    Parameter type is "list[str | Unknown]" (reportUnknownParameterType)
  a.py:6:12 - error: Return type, "list[str | Unknown]", is partially unknown (reportUnknownVariableType)
  a.py:8:13 - information: Type of "identity" is "(x: list[str | Unknown]) -> list[str | Unknown]"
3 errors, 0 warnings, 1 information
```

```
$ pyright b.py
b.py
  b.py:8:13 - information: Type of "identity" is "(x: list[str | Nested]) -> list[str | Nested]"
0 errors, 0 warnings, 1 information
```

```
$ pyright a.py b.py
a.py
  a.py:5:5 - error: Return type, "list[str | Unknown]", is partially unknown (reportUnknownParameterType)
  a.py:5:14 - error: Type of parameter "x" is partially unknown
    Parameter type is "list[str | Unknown]" (reportUnknownParameterType)
  a.py:6:12 - error: Return type, "list[str | Unknown]", is partially unknown (reportUnknownVariableType)
  a.py:8:13 - information: Type of "identity" is "(x: list[str | Unknown]) -> list[str | Unknown]"
b.py
  b.py:5:5 - error: Return type, "list[str | Unknown]", is partially unknown (reportUnknownParameterType)
  b.py:5:14 - error: Type of parameter "x" is partially unknown
    Parameter type is "list[str | Unknown]" (reportUnknownParameterType)
  b.py:6:12 - error: Return type, "list[str | Unknown]", is partially unknown (reportUnknownVariableType)
  b.py:8:13 - information: Type of "identity" is "(x: list[str | Unknown]) -> list[str | Unknown]"
6 errors, 0 warnings, 2 informations
```

```
$ pyright b.py a.py
a.py
  a.py:8:13 - information: Type of "identity" is "(x: list[str | list[StrOrNested]]) -> list[str | list[StrOrNested]]"
b.py
  b.py:8:13 - information: Type of "identity" is "(x: list[str | Nested]) -> list[str | Nested]"
0 errors, 0 warnings, 2 informations
```


### Versions

$ python --version
Python 3.12.3
$ pyright --version
pyright 1.1.404

### Workaround

use of new type alias syntax can avoid the problem:
```py
Nested = list["StrOrNested"]
StrOrNested = str | Nested
type Alias = Nested
```

**Comment by bukzor** [comment]
Reproduces as described for me, in a new directory. I reproduced it with all python versions 3.8-3.13.

```
$ python --version 
Python 3.13.2
```

I did several spot-checks of pyright versions. This seems to be the transition point, but I didn't check exhaustively.

* 1.1.343 -- repro
* 1.1.342 -- no repro

**Comment by erictraut** [comment]
Thanks for the additional info. I was able to repro the problem using the example from @pinterior, and it was indeed a regression introduced in 1.1.343 (almost 8 months ago). This will be fixed in the next release.

**Comment by bukzor** [comment]
I'd love to know any of the details, if you have the time.

**Comment by bukzor** [comment]
I noticed that the fix didn't include a regression test. Would you welcome
a contribution?

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.405
